### PR TITLE
Adding track overlap metrics from `laptrack`

### PR DESCRIPTION
**Pull request by yfukai**
Adding the track purity (TP) and target effectiveness (TE) scores from `laptrack`. See https://github.com/yfukai/laptrack/issues/374 for discussion.

Definitions of the metrics are as follows. They are used in several literatures as the metrics that penalize the false negative and the false positive edges, respectively. 

- TE for a single ground truth track $T^g_j$ is calculated by finding the predicted track $T^p_k$ that overlaps with $T^g_j$ in the largest number of the frames and then dividing the overlap frame counts by the total frame counts for $T^g_j$. The TE for the total dataset is calculated as the mean of TEs for all ground truth tracks, weighted by the length of the tracks.

- TP is defined analogously, with $T^g_j$ and $T^p_k$ being swapped in the definition.

### CODE REVIEW

**Comment by DragaDoncila** [comment]
Thanks for opening this @yfukai - very excited to get this in. I will review as soon as possible!

**Comment by yfukai** [comment]
Thanks @DragaDoncila! The implementation may be sub-optimal, so I'm happy to have feedbacks on this!

File: src/traccuracy/metrics/_track_overlap.py
Comment (cmalinmayor) [inline src/traccuracy/metrics/_track_overlap.py:106]: Is there a reason the total is computed using only the mapped edges instead of `len(reference_tracklet)`? Based on my understanding of the metric, we want to divide by the total length of the reference tracklet, which is different from the mapped edges if there is a false negative node (the `if` on line 97 is false).

**Review by cmalinmayor — COMMENTED** [review]
@yfukai Thank you so much for implementing these metrics! I just took a quick look and the code looks great - easy to follow, well commented, sensible changes to the existing code, has tests! I'll need to look more closely at the efficiency of the implementation because I do have a feeling that there might be a faster way, and I had one question about the correctness, but otherwise it looks fantastic. 

For better or worse, we are doing a pretty active sprint on the library this week, so we will probably introduce some breaking changes, but if we can get this merged with tests before, it will be on us to fix the issues we introduce 😅. Also, if you have any suggestions for improving the overall structure of the library, please add issues!

I will probably follow up with further documentation requests - we are planning to add a page in the docs for each metric as discussed in #77. But I'm happy to merge this first and have a new PR for that documentation.

Closes #112

File: src/traccuracy/metrics/_track_overlap.py
Comment (msschwartz21) [inline src/traccuracy/metrics/_track_overlap.py:37]: In order for the docs to render correctly, can you move this docstring under the class (line 26) itself instead of under init?

File: src/traccuracy/metrics/_track_overlap.py
Comment (msschwartz21) [inline src/traccuracy/metrics/_track_overlap.py:43]: What is this comment referring to? Should it be removed?

File: src/traccuracy/metrics/_track_overlap.py
Comment (cmalinmayor) [reply src/traccuracy/metrics/_track_overlap.py:106]: It would be good to test more cases like this to ensure there aren't any other edge cases we are missing - false negative nodes, false positive nodes, etc.

**Review by msschwartz21 — COMMENTED** [review]
Thank you for contributing! I left two minor comments below. Caroline is correct that some of our open PRs are going to cause all sorts of merge conflicts, but we'll help you get that sorted out.

File: src/traccuracy/metrics/_track_overlap.py
Comment (yfukai) [reply src/traccuracy/metrics/_track_overlap.py:43]: Oh thanks for noticing this, this is the left-off from the copied code. I'll remove this comment!

File: src/traccuracy/metrics/_track_overlap.py
Comment (yfukai) [reply src/traccuracy/metrics/_track_overlap.py:37]: I moved the __init__ docstring to under the class!

File: src/traccuracy/metrics/_track_overlap.py
Comment (yfukai) [reply src/traccuracy/metrics/_track_overlap.py]: Thanks for your comment! Yeah right, I believe it was not correct that I mapped the reference tracklet. I reorganized the code so that it maps the overlapping tracklet. I also added test with false negative nodes, false positive nodes and overlapping detections. I also checked the values of TE and TP swaps with swapping gt and pred tracks.

**Comment by yfukai** [comment]
Thanks for your review @cmalinmayor and @msschwartz21! I believe I could resolve those points. I'm happy with either of merging this now or waiting for API update and resolve conflicts.

File: src/traccuracy/metrics/_track_overlap.py
Comment (cmalinmayor) [inline src/traccuracy/metrics/_track_overlap.py:82]: Can you update this docstring? Thanks!

File: src/traccuracy/metrics/_track_overlap.py
Comment (cmalinmayor) [inline src/traccuracy/metrics/_track_overlap.py:53]: If you make these dictionaries, for example `gt_pred_mapping =  {gt_id: [pred_id, ...]}` and `pred_gt_mapping = {pred_id: [gt_id]}`, you can replace the map_node helper function in the _calc_overlap_score function with a simple dictionary lookup, and increase efficiency by not looping over all pairs to find matched edges.

**Comment by cmalinmayor** [comment]
> Thanks for your review @cmalinmayor and @msschwartz21! I believe I could resolve those points. If this looks OK (maybe after checking the implementation?), I'm happy with either of merging this now or waiting for API update and resolving conflicts.

Thanks for the changes! I had two quick comments, one about a docstring and the other a suggestion to improve efficiency. After we resolve those I think we can merge now and worry about API changes later. Thanks!

File: src/traccuracy/metrics/_track_overlap.py
Comment (yfukai) [reply src/traccuracy/metrics/_track_overlap.py:82]: done!

File: src/traccuracy/metrics/_track_overlap.py
Comment (yfukai) [reply src/traccuracy/metrics/_track_overlap.py]: This is also done! I hope the implementation has become more efficient with this.

**Review by cmalinmayor — APPROVED** [review]
Thanks @yfukai for being so responsive! I'm happy to merge this now, unless @msschwartz21 has any final notes. Whoever updates this for the new API (probably Morgan or I) will request your review to double check any changes, and we will also be following up with a request for documentation soon. Super excited to have these metrics in the library!

**Review by msschwartz21 — APPROVED** [review]

**Review by msschwartz21 — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**
